src/jit/alloc.cpp

```diff
@@ -1,411 +1,389 @@
 // Otherwise, the page specified by temp could be double-freed (VSW 600919).
-void        norls_allocator::nraToss(nraMarkDsc &mark)
+void ArenaAllocator::reset(MarkDescriptor& mark)
 {
-    void    *   last = mark.nmPage;
+    // If the active page hasn't changed, just reset the position into the
+    // page and return.
+    if (m_lastPage == mark.m_page)
+    {
+        m_nextFreeByte = mark.m_next;
+        m_lastFreeByte = mark.m_last;
+        return;
+    }
+
+    // Otherwise, free any new pages that were added.
+    void* last = mark.m_page;
 
-    if  (!last)
+    if (last == nullptr)
     {
-        if  (!nraPageList)
+        if (m_firstPage == nullptr)
+        {
             return;
+        }
 
-        nraFreeNext  = nraPageList->nrpContents;
-        nraFreeLast  = nraPageList->nrpPageSize + (BYTE *)nraPageList;
-
+        m_nextFreeByte = m_firstPage->m_contents;
+        m_lastFreeByte = m_firstPage->m_pageBytes + (BYTE*)m_firstPage;
         return;
     }
 
-    /* Free up all the new pages we've added at the end of the list */
-
-    while (nraPageLast != last)
+    while (m_lastPage != last)
     {
-        norls_pagdesc * temp;
-
-        /* Remove the last page from the end of the list */
+        // Remove the last page from the end of the list
+        PageDescriptor* temp = m_lastPage;
+        m_lastPage = temp->m_previous;
 
-        temp = nraPageLast;
-               nraPageLast = temp->nrpPrevPage;
+        // The new last page has no next page
+        m_lastPage->m_next = nullptr;
 
-        /* The new last page has no 'next' page */
+        freeHostMemory(temp);
+    }
 
-        nraPageLast->nrpNextPage = 0;
+    m_nextFreeByte = mark.m_next;
+    m_lastFreeByte = mark.m_last;
+}
 
-        nraVirtualFree(temp, 0, MEM_RELEASE);
+void* ArenaAllocator::allocateHostMemory(size_t size)
+{
+#if defined(DEBUG)
+    if (!useHostAllocator())
+    {
+#undef GetProcessHeap
+#undef HeapAlloc
+        return ::HeapAlloc(GetProcessHeap(), 0, size);
     }
+    else
+    {
+        return ClrAllocInProcessHeap(0, S_SIZE_T(size));
+    }
+#else // defined(DEBUG)
+    return m_memoryManager->ClrVirtualAlloc(nullptr, size, MEM_COMMIT, PAGE_READWRITE);
+#endif // !defined(DEBUG)
+}
 
-    nraFreeNext = mark.nmNext;
-    nraFreeLast = mark.nmLast;
+void ArenaAllocator::freeHostMemory(void* block)
+{
+#if defined(DEBUG)
+    if (!useHostAllocator())
+    {
+#undef GetProcessHeap
+#undef HeapFree
+        ::HeapFree(GetProcessHeap(), 0, block);
+    }
+    else
+    {
+        ClrFreeInProcessHeap(0, block);
+    }
+#else // defined(DEBUG)
+    m_memoryManager->ClrVirtualFree(block, 0, MEM_RELEASE);
+#endif // !defined(DEBUG)
 }
 
-/*****************************************************************************/
-#ifdef DEBUG
-/*****************************************************************************/
-void    *           norls_allocator::nraAlloc(size_t sz)
+#if defined(DEBUG)
```

Comment:
You should reinforce that this the `DEBUG`-only implementation, and so it has special abilities (fault injection, seeding the allocation with a pattern) that aren't there in the normal version.